notifications/v1/client.go

```diff
@@ -6,90 +6,153 @@ import (
 	ctx, cancel := context.WithCancel(context.Background())
-	c.cancelContext = cancel
-	return c.subscription.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
+	s.cancelContext = cancel
+	err := s.subscription.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
 		var msgData struct {
 			Name     string          `json:"name"`
 			Metadata json.RawMessage `json:"metadata"`
 		}
 		err := json.Unmarshal(msg.Data, &msgData)
 		if err != nil {
-			c.logger.Error("failed to unmarshal notification message", log.Error(err))
+			s.logger.Error("failed to unmarshal notification message", log.Error(err))
 			msg.Nack()
 			return
 		}
 
-		err = c.handleReceive(handler, msgData.Name, msgData.Metadata)
-		if err != nil {
-			c.logger.Error("failed to process notification message", log.Error(err))
+		err = s.handleReceive(msgData.Name, msgData.Metadata)
+		if err == nil {
+			msg.Ack()
+		} else {
+			s.logger.Error("failed to process notification message", log.Error(err))
 			msg.Nack()
-			return
 		}
-		msg.Ack()
 	})
+	if err != nil {
+		s.logger.Error("failed to receive notifications", log.Error(err))
+		return
+	}
+}
+
+func (r *subscriber) Stop() {
+	if !r.state.CompareAndSwap(stateStarted, stateStopped) {
+		r.logger.Error("failed to stop subscriber", log.Error(errors.New(`not in "started" state`)))
+		return
+	}
+
+	r.cancelContext()
+	r.logger.Info("subscriber stopped")
 }
 
-func (c *Client) handleReceive(handler *ReceiveHandlers, name string, metadata json.RawMessage) error {
+// SubscriberHandlers is a collection of subscription handlers for each type of
+// SAMS notifications. If the handler of a notification is nil, the notification
+// will be acknowledged automatically without any processing.
+//
+// If a handler returns an error, the notification will be unacknowledged and
+// retried later.
+type SubscriberHandlers struct {
+	// OnUserDeleted is called when a "UserDeleted" notification is received.
+	OnUserDeleted func(data *UserDeletedData) error
+}
+
+type ReceiveSettings = pubsub.ReceiveSettings
+
+var DefaultReceiveSettings = pubsub.DefaultReceiveSettings
+
+func (r *subscriber) handleReceive(name string, metadata json.RawMessage) error {
 	switch name {
 	case nameUserDeleted:
+		if r.handlers.OnUserDeleted == nil {
+			return nil
+		}
+
 		var data UserDeletedData
-		err := json.Unmarshal(metadata, &data)
-		if err != nil {
+		if err := json.Unmarshal(metadata, &data); err != nil {
 			return errors.Wrap(err, "unmarshal metadata")
 		}
-		if handler.OnUserDeleted != nil {
-			return handler.OnUserDeleted(&data)
-		}
+		return r.handlers.OnUserDeleted(&data)
 	default:
-		c.logger.Warn("acknowledging unknown notification name", log.String("name", name))
+		r.logger.Warn("acknowledging unknown notification name", log.String("name", name))
 	}
 	return nil
 }
 
-// Close stops the client from receiving notifications.
-func (c *Client) Close() {
-	c.cancelContext()
-}
+const (
```

Comment:
I, being brain damaged by the functional programming disease, an morally obligated to point out that we can TOTALLY USE A TYPE HERE.

```go
type subscriberState string
const (
	stateIdle    = "idle"
	stateStarted = "started"
	stateStopped = "stopped"
)
```

FAQ answers:

1. Yes, this isn't actually of much benefit.
2. Yes, this will require you to add `string(...)` casts in various places.
3. Yes, as a card carrying functional programming wonk, I will have my membership status revoked unless I point out any and every opportunity to lean on the type checker to catch things like this.
4. No, I don't care if you ignore this suggestion. In fact, I would be annoyed if someone asked me to. But see point number 3.